ERC20::ERC1155 Exchange

ScreamingHawk opened the pull request:
I've had a go at creating the ERC20::ERC1155 Exchange as per #32. 
It's essentially a clone of the existing Factory methods with small changes and a new `buyTokens` method for purchasing with the ERC20 token. 
All the existing tests have been copied across and updated as required in order to pass. 

Let me know if this is on track or if there are any changes required.

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
Should require that this returns true since not all ERC-20 transfer implementation will throw on error.

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
Should require that this returns true since not all ERC-20 transfer implementation will throw on error.

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
Should require that this returns true since not all ERC-20 transfer implementation will throw on error.

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
Should require that this returns true since not all ERC-20 transfer implementation will throw on error.

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
Should require that this returns true since not all ERC-20 transfer implementation will throw on error.

PhABC commented on src/contracts/exchange/NiftyswapFactory20.sol:
```diff
@@ -0,0 +1,38 @@
+   * @param _currency   The address of the ERC-20 token contract
+   */
+  function createExchange(address _token, address _currency) public override {
+    require(tokensToExchange[_token][_currency] == address(0x0), "NiftyswapFactory#createExchange: EXCHANGE_ALREADY_CREATED");
```
error message should say `NiftyswapFactory20#...`

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
error message should say `NiftyswapExchange20#...`

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
error message should say `NiftyswapExchange20#...`

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
error message should say `NiftyswapExchange20#...`

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
error message should say `NiftyswapExchange20#...`

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
error message should say `NiftyswapExchange20#...`

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
error message should say `NiftyswapExchange20#...`

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
error message should say `NiftyswapExchange20#...`

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
error message should say `NiftyswapExchange20#...`

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
error message should say `NiftyswapExchange20#...`

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
error message should say `NiftyswapExchange20#...`

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
error message should say `NiftyswapExchange20#...`

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
error message should say `NiftyswapExchange20#...`

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
error message should say `NiftyswapExchange20#...`

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
error message should say `NiftyswapExchange20#...`

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
error message should say `NiftyswapExchange20#...`

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
error message should say `NiftyswapExchange20#...`

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
error message should say `NiftyswapExchange20#...`

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
error message should say `NiftyswapExchange20#...`

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
error message should say `NiftyswapExchange20#...`

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
error message should say `NiftyswapExchange20#...`

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
error message should say `NiftyswapExchange20#...`

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
error message should say `NiftyswapExchange20#...`

PhABC commented on src/contracts/exchange/NiftyswapExchange20.sol:
```diff
@@ -0,0 +1,812 @@
+ * with an ERC-20 based token used as a currency instead of Ether.
+ *
+ * See https://github.com/0xsequence/erc20-meta-token for a generalized
+ * ERC-20 => ERC-1155 token wrapper
```
You can delete these two lines since we don't need to wrap ERC-20 into ERC-1155 with this contract.

PhABC commented:
Thanks for the PR @ScreamingHawk! It's looking good, just commented on minor things. @Agusx1211 will also review it and then we would send for security audit.

ScreamingHawk commented:
Thanks for the thorough review @PhABC ! I've gone through and updated all the error messaging (in the tests as well) and have required a true response for all the transfer and transferFrom. And fixed the doc comments. 
Do you want me to go through and resolve all those review threads or shall I leave them for you to double check?

Agusx1211 commented on src/contracts/exchange/NiftyswapExchange20.sol:
Some ERC20 tokens don't return true on `transfer` but some do, for maximum compatibility with tokens that don't fully comply with the standard we should use something like the Uniswap TransferHelper.

https://github.com/Uniswap/uniswap-lib/blob/master/contracts/libraries/TransferHelper.sol

Agusx1211 replied:
This applies to all the ERC20 transfers

ScreamingHawk commented:
Thanks for the review @Agusx1211 ! 
I've updated the package to include uniswap's TransferHelper and updated all the references as requested.

Agusx1211 commented on src/contracts/exchange/NiftyswapExchange20.sol:
```diff
@@ -0,0 +1,810 @@
+  // Variables
+  IERC1155 internal token;                        // address of the ERC-1155 token contract
+  address internal currency;                      // address of the ERC-20 currency used for exchange
+  address internal factory;                       // address for the factory that created this contract
```
```
  IERC1155 internal token;                        // address of the ERC-1155 token contract
  address internal currency;                      // address of the ERC-20 currency used for exchange
  address internal factory;      
```

Are there any reasons for not defining these variables as `immutable`?

ScreamingHawk replied:
Only because they weren't in `NiftwapExchange.sol`. Will update

Agusx1211 commented on src/contracts/exchange/NiftyswapExchange20.sol:
```diff
@@ -0,0 +1,810 @@
+   */
+  constructor(address _tokenAddr, address _currencyAddr) public {
+    require(
+      address(_tokenAddr) != address(0) && _currencyAddr != address(0),
```
`_tokenAddr` is already of address type, no need to cast it to address

ScreamingHawk replied:
Removed casting

Agusx1211 commented on src/contracts/exchange/NiftyswapExchange20.sol:
I am fairly sure this array is never used, I think we can replace this line and the following one with just:

`uint256[] memory tokenReserves = _getTokenReserves(_tokenIds);`

ScreamingHawk replied:
Sorry I had this in for debugging. Will remove

Agusx1211 commented on src/contracts/exchange/NiftyswapExchange20.sol:
Same as before, I don't think this array has to be declared in a different line.

ScreamingHawk replied:
Will update

Agusx1211 commented on src/contracts/exchange/NiftyswapExchange20.sol:
I think `tokenReserves` doesn't have to be created before calling `_getTokenReserves`

ScreamingHawk replied:
Will update

Agusx1211 commented on src/contracts/exchange/NiftyswapExchange20.sol:
Same as before with `tokenReserves`

ScreamingHawk replied:
Will update

Agusx1211 commented:
Added some extra comments

ScreamingHawk commented:
Sorry for the delay! 
Updated as per all comments. Rebased onto master to avoid that merge conflict. Regenerated the files. 
Let me know if there is anything else! 😃
